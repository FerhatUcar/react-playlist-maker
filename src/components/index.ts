/**
 * @file Automatically generated by barrelsby.
 */

export * from "./Alert/Alert";
export * from "./Alert/styled";
export * from "./Button/Button";
export * from "./Button/style";
export * from "./List/List";
export * from "./List/style";
export * from "./Loader/Loader";
export * from "./Loader/styled";
export * from "./Search/Search";
export * from "./Search/style";
