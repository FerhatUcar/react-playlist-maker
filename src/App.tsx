import React, { useCallback, useEffect, useState } from "react";
import "./assets/scss/App.scss";
import {
  Container,
  Playlist,
  BoxHeader,
  FullList,
  Layout,
} from "./style/globalStyle";
import { List, Search, Loader, Alert } from "./components";
import { EventInputType, RED_COLOR } from "./style/config";
import { Header, Menu } from "./blocks";
import { globalFilter } from "./helpers/globalFilter";
import { Song } from "./types/song";
import { deletePlaylistSong, getPlaylist, getSongs, postSong } from "./songs";
import { faHeart } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const App = (): JSX.Element => {
  const [expand, setExpand] = useState(false);
  const [search, setSearch] = useState("");
  const [songs, setSongs] = useState<Song[]>([]);
  const [playlist, setPlaylist] = useState<Song[]>([]);
  const [playlistComponent, setPlaylistComponent] = useState(true);
  const [loading, setLoading] = useState(false);
  const [alreadyAdded, setAlreadyAdded] = useState(false);
  const [alreadyExist, setAlreadyExist] = useState(false);

  useEffect(() => {
    setLoading(true);

    getSongs().then((data: Song[]) => {
      setSongs(data.slice(0, 5));
      setLoading(false);
    });
  }, [setLoading]);

  useEffect(() => {
    getPlaylist().then((data: Song[]) => setPlaylist(data));
  }, []);

  const filter = useCallback(
    (e: EventInputType) => {
      setSearch(e.target.value);
    },
    [setSearch]
  );

  const handleDelete = useCallback(
    ({ id }: Song) => {
      deletePlaylistSong(id);

      const newPlaylistSongs = playlist.filter((song) => song.id !== id);
      setPlaylist(newPlaylistSongs);
    },
    [playlist]
  );

  const addSongToPlaylist = useCallback(
    (item: Song) => {
      for (let i = 0; i < playlist.length; i++) {
        if ([item.id].includes(playlist[i].id)) {
          return setAlreadyAdded(true);
        }
      }

      postSong(item);
      setPlaylist((playlist) => [...playlist, item]);

      const newSongs = songs.filter((song) => song.id !== item.id);
      setSongs(newSongs);
      setAlreadyAdded(false);
    },
    [setPlaylist, setSongs, songs, playlist, setAlreadyAdded]
  );

  const closeAlert = useCallback(() => {
    setAlreadyAdded(false);
    setAlreadyExist(false);
  }, [setAlreadyAdded, setAlreadyExist]);

  const openMenu = useCallback(() => setExpand(!expand), [setExpand, expand]);

  const fetchMoreSongs = () => {
    setTimeout(() => {
      getSongs().then((data: Song[]) => {
        setSongs(data.slice(0, songs.length + 5));
      });
    }, 1000);
  };

  const addPlaylist = useCallback(() => {
    if (playlistComponent) {
      setExpand(false);
      return setAlreadyExist(true);
    }

    setPlaylistComponent(true);
  }, [setPlaylistComponent, setAlreadyExist, playlistComponent, setExpand]);

  return (
    <>
      <Layout expand={expand}>
        <Header openMenu={openMenu} expanded={expand} />
        <Container>
          {alreadyAdded && (
            <Alert
              text="Song already added to your playlist!"
              closeAlert={closeAlert}
            />
          )}
          {alreadyExist && (
            <Alert
              text="Playlist has already been created"
              closeAlert={closeAlert}
            />
          )}
          {playlistComponent && (
            <Playlist>
              <BoxHeader>
                <div>
                  <span>Playlist</span>
                  <FontAwesomeIcon color={RED_COLOR} icon={faHeart} />
                </div>
              </BoxHeader>
              <List
                infinite={false}
                onClick={handleDelete}
                data={playlist}
                option="delete"
              />
            </Playlist>
          )}

          <FullList>
            <BoxHeader>
              <span>Full list</span>
              <Search filter={filter} />
            </BoxHeader>
            {loading ? (
              <Loader />
            ) : (
              songs && (
                <List
                  infinite
                  fetchMoreData={fetchMoreSongs}
                  hasMoreData
                  option="add"
                  data={globalFilter(songs, search).map((item) => item)}
                  onClick={addSongToPlaylist}
                />
              )
            )}
          </FullList>
        </Container>
      </Layout>
      <Menu playlist={addPlaylist} expand={expand} />
    </>
  );
};

export default App;
