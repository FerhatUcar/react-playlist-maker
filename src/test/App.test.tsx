import React from "react";
import { render, screen } from "@testing-library/react";
import App from "../App";
import Enzyme from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { List } from "../components";
import renderer from 'react-test-renderer';

Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {
  it("renders Rockstars App title in header", () => {
    render(<App />);
    const headerTitle = screen.getByText(/Rockstars App/i);
    expect(headerTitle).toBeInTheDocument();
  });

  it('snapshot renders', () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the songs list', () => {
    const onClick = jest.fn();
    const { getByTestId } = render(<List onClick={onClick} />);
    const element = getByTestId('songs-list')
    expect(element).toBeInTheDocument();
  });
})
