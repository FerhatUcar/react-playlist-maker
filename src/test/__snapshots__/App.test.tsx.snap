// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders 1`] = `
Array [
  <div
    className="sc-jSFjdj dmiDui"
  >
    <div
      className="sc-iqAclL dvKyuS"
    >
      <div
        className="sc-crzoAE eagBrb"
      >
        <img
          alt="logo"
          className="sc-gtsrHT gKttTv"
          src="logo.png"
        />
        <h1>
          Rockstars app
        </h1>
      </div>
      <div
        className="sc-dIsUp cAtrdt"
      >
        <button
          className="sc-bqGGPW jMXrbk"
          onClick={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-hamburger fa-w-16 "
            data-icon="hamburger"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M464 256H48a48 48 0 0 0 0 96h416a48 48 0 0 0 0-96zm16 128H32a16 16 0 0 0-16 16v16a64 64 0 0 0 64 64h352a64 64 0 0 0 64-64v-16a16 16 0 0 0-16-16zM58.64 224h394.72c34.57 0 54.62-43.9 34.82-75.88C448 83.2 359.55 32.1 256 32c-103.54.1-192 51.2-232.18 116.11C4 180.09 24.07 224 58.64 224zM384 112a16 16 0 1 1-16 16 16 16 0 0 1 16-16zM256 80a16 16 0 1 1-16 16 16 16 0 0 1 16-16zm-128 32a16 16 0 1 1-16 16 16 16 0 0 1 16-16z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      className="sc-bdnxRM hxCrGu"
    >
      <div
        className="sc-dlnjwi kjAgwx"
      >
        <div
          className="sc-eCApnc kMjsgI"
        >
          <div>
            <span>
              Playlist
            </span>
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-heart fa-w-16 "
              color="#ea3100"
              data-icon="heart"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </div>
        </div>
        <ul
          className="sc-fujyAs jrbMWL"
          id=""
        />
      </div>
      <div
        className="sc-hKFxyN jXaqPd"
      >
        <div
          className="sc-eCApnc kMjsgI"
        >
          <span>
            Full list
          </span>
          <input
            className="sc-jrsJWt eRzbuS"
            onChange={[Function]}
            placeholder="Search artist.."
            type="input"
          />
        </div>
        <ul
          className="sc-fujyAs jrbMWL"
          id="scrollableDiv"
        >
          <div
            className="infinite-scroll-component__outerdiv"
            style={Object {}}
          >
            <div
              className="infinite-scroll-component "
              style={
                Object {
                  "WebkitOverflowScrolling": "touch",
                  "height": "auto",
                  "overflow": "auto",
                }
              }
            >
              <div
                className="sc-pNWdM iNdOsy"
              />
            </div>
          </div>
        </ul>
      </div>
    </div>
  </div>,
  <div
    className="sc-kEqXSa iCPckk"
  >
    <h3>
      Menu
    </h3>
    <hr />
    <button
      onClick={[Function]}
    >
      Add playlist
    </button>
  </div>,
]
`;
